sam.c

```diff
@@ -2493,19 +2614,19 @@ static int sam_parse_B_vals(char type, uint32_t n, char *in, char **end,
         if (!overflow) {
             if (min < 0) {
                 if (min >= INT8_MIN && max <= INT8_MAX) {
-                    return sam_parse_B_vals('c', n, in, end, r, b);
+                    return sam_parse_B_vals('c', n, in, end, b);
                 } else if (min >= INT16_MIN && max <= INT16_MAX) {
-                    return sam_parse_B_vals('s', n, in, end, r, b);
+                    return sam_parse_B_vals('s', n, in, end, b);
```

Comment:
It would be more efficient to pass `nused` for the second attempt.  Increasing the data size may trigger a reallocation, but at this point we know how many slots are really needed so we can ask for the exact size required.

Reply:
It shouldn't matter, but actually I see both are wrong.  Our `possibly_expand_bam_data` is always growing, unlike the kstring equivalent which is asking for a size capable of holding at least 'x' bytes.  Hence whatever we pass in to the next loop when we recurse will be what we have plus a load more.  That's wrong, as we don't need any more at all.

The whole recursion thing is a bit of a mess frankly.  I need to rethink it.